File: dbhandler.go
package main

import (
    _ "github.com/mattn/go-sqlite3"
    "database/sql"
    "fmt"
    "time"
)

/* -- Constants -- */
const TABLE_EXPOSES = `create table if not exists exposes 
						(id varchar(20), name text, price_warm int, price_cold int, last_seen int, 
						first_seen int, care varchar(1), pets varchar(1),
						zipcode varchar(10), city varchar(30), dist_work int,
						kausion int, url text, collected varchar(1), rooms int, size int,
						online varchar(1));`

const DB_FILE_NAME = "./list.db"

const TABLE_EXPOSES_NO = "N"
const TABLE_EXPOSES_YES = "Y"

type Expose struct {
	id string
	name string
	price_warm int
	price_cold int
	last_seen int
	first_seen int
	care string
	pets string 
	zipcode string
	city string
	dist_work int
	kausion int
	url string
	collected string
	rooms int
	size int
	online string
}


type DBHandler struct {
	db *sql.DB
}


/**
 * Opens a connection to the databasefile, creates one if it does not exits
 * 
 * @name Init
 */
func (this *DBHandler) Init() {
	var derr error
	this.db, derr = sql.Open("sqlite3", DB_FILE_NAME)
	if derr != nil {
		fmt.Println(derr)
	}
	this.createNewTable(TABLE_EXPOSES)
}


/**
 * Closes the connection to the databasefile
 * 
 * @name Close
 */
func (this *DBHandler) Close() {
	this.db.Close()
}


/**
 * Stores a wordlist to the databasefile
 * 
 * @name StoreWordlist
 */
func (this *DBHandler) StoreExpose(expose *Expose) {
	if ! this.isExposeUniqe(expose.id) {
		fmt.Printf("Expose (%s) already exists, updating last seen.\n", expose.id)
		exposeObj := this.GetExposeById(expose.id)
		exposeObj.last_seen = int(time.Now().Unix())
		this.UpdateExpose(&exposeObj)
		return
	}

	tx, err := this.db.Begin()
	if err != nil {
		fmt.Println(err)
	}
	stmt, err := tx.Prepare("insert into exposes(id, name, price_warm, price_cold, first_seen, last_seen, care, pets, zipcode, city, dist_work, kausion, url, collected, rooms, size, online) values(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	_, err = stmt.Exec(expose.id, expose.name, expose.price_warm, expose.price_cold, expose.first_seen, expose.last_seen, TABLE_EXPOSES_YES, expose.pets, expose.zipcode, expose.city, expose.dist_work, expose.kausion, expose.url, TABLE_EXPOSES_NO, expose.rooms, expose.size, expose.online)
	if err != nil {
		fmt.Println(err)
	}
	tx.Commit()
}


/**
 * Update a expose in the databasefile
 * 
 * @name UpdateExpose
 */
func (this *DBHandler) UpdateExpose(expose *Expose) {
	tx, err := this.db.Begin()
	if err != nil {
		fmt.Println(err)
	}
	stmt, err := tx.Prepare("update exposes set name=?, price_warm=?, price_cold=?, first_seen=?, last_seen=?, care=?, pets=?, zipcode=?, city=?, dist_work=?, kausion=?, url=?, collected=?, rooms=?, size=?, online=? where id=?")
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	_, err = stmt.Exec(expose.name, expose.price_warm, expose.price_cold, expose.first_seen, expose.last_seen, expose.care, expose.pets, expose.zipcode, expose.city, expose.dist_work, expose.kausion, expose.url, expose.collected, expose.rooms, expose.size, expose.online, expose.id)
	if err != nil {
		fmt.Println(err)
	}
	tx.Commit()
}


func (this *DBHandler) GetAllNonCollectedExposes() []Expose {
	listexpose := []Expose{}
	rows, err := this.db.Query("select * from exposes where collected = '"+TABLE_EXPOSES_NO+"'")
	if err != nil {
		fmt.Println(err)
	}
	defer rows.Close()
	for rows.Next() {
		var expose Expose
		err = rows.Scan(&expose.id, &expose.name, &expose.price_warm, &expose.price_cold, &expose.first_seen, &expose.last_seen, &expose.care, &expose.pets, &expose.zipcode, &expose.city, &expose.dist_work, &expose.kausion, &expose.url, &expose.collected, &expose.rooms, &expose.size, &expose.online)
		if err != nil {
			fmt.Println(err)
		}
		listexpose = append(listexpose, expose)
	}
	err = rows.Err()
	if err != nil {
		fmt.Println(err)
	}
	return listexpose
}


func (this *DBHandler) GetExposeById(id string) Expose {
	stmt, err := this.db.Prepare("select * from exposes where id = ?")
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	var expose Expose
	err = stmt.QueryRow(id).Scan(&expose.id, &expose.name, &expose.price_warm, &expose.price_cold, &expose.first_seen, &expose.last_seen, &expose.care, &expose.pets, &expose.zipcode, &expose.city, &expose.dist_work, &expose.kausion, &expose.url, &expose.collected, &expose.rooms, &expose.size, &expose.online)
	if err != nil {
		fmt.Println(err)
	}
	return expose
}


/**
 * Runs a table script on the database file.
 * 
 * @name createNewTable
 */
func (this *DBHandler) createNewTable(tablescript string) {
	_, err := this.db.Exec(tablescript)
	if err != nil {
		fmt.Printf("%q: %s\n", err, tablescript)
		return
	}
}



/**
 * Checks if the given expose id is uniqe
 * 
 * @name isExposeUniqe
 * @return bool
 */
func (this *DBHandler) isExposeUniqe(id string) bool {
	stmt, err := this.db.Prepare("select count(*) from exposes where id = ?")
	if err != nil {
		fmt.Println(err)
	}
	defer stmt.Close()
	var count int
	err = stmt.QueryRow(id).Scan(&count)
	if err != nil {
		fmt.Println(err)
	}
	if count > 0 {
		return false
	}
	return true
}

File: dbhandler.php
<?php

class DBHandler
{
	const TABLE_EXPOSES_TRUE = 'Y';
	const TABLE_EXPOSES_FALSE = 'N';
	const TABLE_EXPOSES_MAYBE = 'M';

	private $db;
	private $TABLE_EXPOSES = 'create table if not exists exposes 
						(id varchar(20), name text, price_warm int, price_cold int, last_seen int, 
						first_seen int, score int, pets varchar(1),
						zipcode varchar(10), city varchar(50), dist_work int,
						kausion int, url text, collected varchar(1), rooms int, size int,
						online varchar(1), next_check int, floor varchar(50), access varchar(50),
						street varchar(50), next_mail int);';
	private $CONFIG;

	function __construct($config)
	{
		$this->CONFIG = $config;
		$this->db = new SQLite3($this->CONFIG['db_file_path']);
		$this->db->query($this->TABLE_EXPOSES);
	}


	function Store_Expose($expose)
	{
		$query = 'insert into exposes(';
		$itr = 0;
		foreach ($expose as $key => $value) 
		{
			$query .= $key;
			$query .= ($itr < count($expose) -1 ? ', ' : ') ');
			$itr++;
		}
		$query .= 'values(';
		$itr = 0;
		foreach ($expose as $key => $value) 
		{
			$query .= ':'.$key;
			$query .= ($itr < count($expose) -1 ? ', ' : ')');
			$itr++;
		}
		$stmt = $this->db->prepare($query);
		foreach ($expose as $key => $value) 
		{
			$stmt->bindValue(':'.$key, $value);
		}
		$result = $stmt->execute();	
	}


	function Get_Exposes_For_Collection()
	{
		$stmt = $this->db->prepare('select * from exposes where next_check < :now and online != :online');
		$stmt->bindValue(':now', time());
		$stmt->bindValue(':online', DBHandler::TABLE_EXPOSES_FALSE);
		$result = $stmt->execute();		
		$list = array();
		while ($row = $result->fetchArray()) {
			$list[] = $this->_clean_expose_result($row);
		}
		return $list;
	}


	function Get_Expose_By_Id($id)
	{
		$stmt = $this->db->prepare('select * from exposes where id = :id');
		$stmt->bindValue(':id', $id);
		$result = $stmt->execute();		
		$list = array();
		while ($row = $result->fetchArray()) {
            $list[] = $this->_clean_expose_result($row);
		}
		return reset($list);
	}


    function Get_Exposes_For_Scorecalc()
    {
        $stmt = $this->db->prepare('select * from exposes where online != :online');
        $stmt->bindValue(':online', DBHandler::TABLE_EXPOSES_FALSE);
        $result = $stmt->execute();
        $list = array();
        while ($row = $result->fetchArray()) {
            $list[] = $this->_clean_expose_result($row);
        }
        return $list;
    }


    function Get_Exposes_By_Score($score)
    {
        $stmt = $this->db->prepare('select * from exposes where score > :score and next_mail < :now');
        $stmt->bindValue(':score', $score);
        $stmt->bindValue(':now', time());
        $result = $stmt->execute();
        $list = array();
        while ($row = $result->fetchArray()) {
            $list[] = $this->_clean_expose_result($row);
        }
        return $list;
    }


	function Update_Expose($expose) 
	{
		$query = 'update exposes set ';
		$itr = 0;
		foreach ($expose as $key => $value) 
		{
			if ($key != 'id')
			{
				$query .= $key.'=:'.$key;
				$query .= ($itr < count($expose) -1 ? ', ' : ' ');
			}
			$itr++;
		}
		$query .= 'where id=:id';
		$stmt = $this->db->prepare($query);
		foreach ($expose as $key => $value) 
		{
			$stmt->bindValue(':'.$key, $value);
		}
		$result = $stmt->execute();	
	}


	function Get_default_expose()
	{
		return 	array(
			'id' => '',
			'name' => '',
			'price_warm' => 0,
			'price_cold' => 0,
			'first_seen' => 0,
			'last_seen' => 0,
			'score' => 0,
			'pets' => '',
			'zipcode' => '',
			'city' => '',
			'dist_work' => 0,
			'kausion' => 0,
			'url' => '',
			'collected' => '',
			'rooms' => 0,
			'size' => 0,
			'online' => '',
			'next_check' => 0,
			'floor' => '',
			'access' => '',
			'next_mail' => 0
		);
	}


	function _clean_expose_result($expose)
	{
		$default_expose = $this->Get_default_expose();
		$result = $this->Get_default_expose();
		foreach ($default_expose as $key => $value)
		{
			$result[$key] = $expose[$key];
		}
		return $result;
	}
}


?>
